libcudacxx/include/cuda/std/detail/libcxx/include/__utility/pair.h

```diff
@@ -69,707 +70,704 @@
+                              && !(_LIBCUDACXX_TRAIT(is_copy_assignable, _TestSynthesizeAssignment<_T1>)
+                                   && _LIBCUDACXX_TRAIT(is_copy_assignable, _TestSynthesizeAssignment<_T2>)))>
+{};
+
+// base class to ensure `is_trivially_copyable` when possible
+template <class _T1, class _T2, bool = __must_synthesize_assignment<_T1, _T2>::value>
+struct __pair_base
+{
+  _T1 first;
+  _T2 second;
+
+  template <class _Constraints                                                 = __pair_constraints<_T1, _T2>,
+            __enable_if_t<_Constraints::__implicit_default_constructible, int> = 0>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr __pair_base() noexcept(
+    _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T1)
+    && _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T2))
+      : first()
+      , second()
+  {}
+
+  template <class _Constraints                                                 = __pair_constraints<_T1, _T2>,
+            __enable_if_t<_Constraints::__explicit_default_constructible, int> = 0>
+  _LIBCUDACXX_INLINE_VISIBILITY explicit constexpr __pair_base() noexcept(
+    _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T1)
+    && _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T2))
+      : first()
+      , second()
+  {}
+
+  template <class _U1, class _U2>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr __pair_base(_U1&& __t1, _U2&& __t2) noexcept(
+    _LIBCUDACXX_TRAIT(is_nothrow_constructible, _T1, _U1) && _LIBCUDACXX_TRAIT(is_nothrow_constructible, _T2, _U2))
+      : first(_CUDA_VSTD::forward<_U1>(__t1))
+      , second(_CUDA_VSTD::forward<_U2>(__t2))
+  {}
+
+protected:
+  template <class... _Args1, class... _Args2, size_t... _I1, size_t... _I2>
+  _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX17 __pair_base(
+    piecewise_construct_t,
+    tuple<_Args1...>& __first_args,
+    tuple<_Args2...>& __second_args,
+    __tuple_indices<_I1...>,
+    __tuple_indices<_I2...>);
+};
 
-    _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX17
-    pair& operator=(__conditional_t<
-                        is_move_assignable<first_type>::value &&
-                        is_move_assignable<second_type>::value,
-                    pair, __nat>&& __p)
-        noexcept(is_nothrow_move_assignable<first_type>::value &&
-                   is_nothrow_move_assignable<second_type>::value)
-    {
-        first = _CUDA_VSTD::forward<first_type>(__p.first);
-        second = _CUDA_VSTD::forward<second_type>(__p.second);
-        return *this;
-    }
+// We need to ensure that a reference type, which would inhibit the implicit copy assignment still works
+template <class _T1, class _T2>
+struct __pair_base<_T1, _T2, true>
+{
+  _T1 first;
+  _T2 second;
+
+  template <class _Constraints                                                 = __pair_constraints<_T1, _T2>,
```

Comment:
question: symbols are quite different now:
```
_ZN4cuda3std3__44pairI1AS3_EC2IS3_S3_NS1_18__pair_constraintsIS3_S3_E15__constructibleIS3_S3_EETnNS1_9enable_ifIXsrT1_24__implicit_constructibleEiE4typeELi0EEEOT_OT0_
```
vs
```
_ZN4cuda3std3__44pairI1AS3_EC2IS3_S3_TnPNS1_9enable_ifIXsr10_CheckArgs17__enable_implicitIT_T0_EE5valueEvE4typeELPv0EEEOS7_OS8_
```
Do we need to bump ABI version?

Reply:
Yeah, I _think_ this is fine. Code that was compiled with the old version will have distinct mangled names from the new version, so they can easily coexist in the same program. As @miscco said, at worst this is a linker error forcing recompilation.

When we do a release, however, this should probably be spelled out as a change potentially requiring recompilation in the release notes.